src/Microsoft.TemplateEngine.Cli/Commands/create/InstantiateCommand.TabCompletion.cs

```diff
@@ -4,23 +4,30 @@
 using System.CommandLine.Completions;
 using System.CommandLine.Parsing;
 using Microsoft.TemplateEngine.Abstractions;
+using Microsoft.TemplateEngine.Edge;
 using Microsoft.TemplateEngine.Edge.Settings;
 
 namespace Microsoft.TemplateEngine.Cli.Commands
 {
     internal partial class InstantiateCommand : BaseCommand<InstantiateCommandArgs>
     {
+        private const int ConstraintEvaluationTimeout = 1000;
+
         internal static IEnumerable<CompletionItem> GetTemplateNameCompletions(string? tempalteName, IEnumerable<TemplateGroup> templateGroups, IEngineEnvironmentSettings environmentSettings)
         {
+            TemplateConstraintManager constraintManager = new TemplateConstraintManager(environmentSettings);
             if (string.IsNullOrWhiteSpace(tempalteName))
             {
-                return templateGroups
+                return GetAllowedTemplateGroups(constraintManager, templateGroups)
                     .SelectMany(g => g.ShortNames, (g, shortName) => new CompletionItem(shortName, documentation: g.Description))
                     .Distinct()
                     .OrderBy(c => c.Label, StringComparer.OrdinalIgnoreCase)
                     .ToArray();
             }
-            return templateGroups
+
+            var matchingTemplateGroups = templateGroups.Where(t => t.ShortNames.Any(sn => sn.StartsWith(tempalteName)));
```

Comment:
Do we need to consider casing here?

Reply:
we were not considering it before, but I think it may be case-insensitive.